File: packages/core/lib/models/resource.dart
/// Data model for resources returned by the API.
///
/// Maps backend ResourceOut / ResourceDetail schemas:
///   GET /api/v1/resources (list)
///   GET /api/v1/resources/{id} (detail, adds signed_pdf_url)
///
/// See backend/app/schemas/resource.py.

class Resource {
  final String id;
  final String title;
  final String category; // phonics | word_card | recommended
  final String? pdfUrl;
  final int sortOrder;
  final bool isActive;
  final String createdAt;
  final String updatedAt;

  const Resource({
    required this.id,
    required this.title,
    required this.category,
    this.pdfUrl,
    this.sortOrder = 0,
    this.isActive = true,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Resource.fromJson(Map<String, dynamic> json) => Resource(
        id: json['id'] as String,
        title: json['title'] as String,
        category: json['category'] as String,
        pdfUrl: json['pdf_url'] as String?,
        sortOrder: json['sort_order'] as int? ?? 0,
        isActive: json['is_active'] as bool? ?? true,
        createdAt: json['created_at'] as String? ?? '',
        updatedAt: json['updated_at'] as String? ?? '',
      );

  /// Chinese label for the category.
  String get categoryLabel {
    switch (category) {
      case 'phonics':
        return '自然拼读';
      case 'word_card':
        return '单词卡';
      case 'recommended':
        return '推荐资源';
      default:
        return category;
    }
  }

  /// Whether this resource has a PDF available for preview.
  bool get canPreview => pdfUrl != null && pdfUrl!.isNotEmpty;
}

/// Resource detail with signed PDF URL (from GET /resources/{id}).
class ResourceDetail extends Resource {
  final String? signedPdfUrl;

  const ResourceDetail({
    required super.id,
    required super.title,
    required super.category,
    super.pdfUrl,
    super.sortOrder,
    super.isActive,
    required super.createdAt,
    required super.updatedAt,
    this.signedPdfUrl,
  });

  factory ResourceDetail.fromJson(Map<String, dynamic> json) =>
      ResourceDetail(
        id: json['id'] as String,
        title: json['title'] as String,
        category: json['category'] as String,
        pdfUrl: json['pdf_url'] as String?,
        sortOrder: json['sort_order'] as int? ?? 0,
        isActive: json['is_active'] as bool? ?? true,
        createdAt: json['created_at'] as String? ?? '',
        updatedAt: json['updated_at'] as String? ?? '',
        signedPdfUrl: json['signed_pdf_url'] as String?,
      );
}

/// Paginated response for resource list.
class PaginatedResources {
  final List<Resource> items;
  final int total;
  final int page;
  final int pageSize;

  const PaginatedResources({
    required this.items,
    required this.total,
    required this.page,
    required this.pageSize,
  });

  factory PaginatedResources.fromJson(Map<String, dynamic> json) =>
      PaginatedResources(
        items: (json['items'] as List)
            .map((e) => Resource.fromJson(e as Map<String, dynamic>))
            .toList(),
        total: json['total'] as int? ?? 0,
        page: json['page'] as int? ?? 1,
        pageSize: json['page_size'] as int? ?? 20,
      );
}

File: backend/app/schemas/resource.py
"""Resource module Pydantic schemas.

Storage migration (2026-06):
- category: opened from hardcoded regex to free string (max 50 chars)
- pdf_url: now stores logical path like "resources/{category}/{id}.pdf"
- UploadPdfOut.storage_path: logical path, not Supabase path
"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Resource CRUD — admin
# ---------------------------------------------------------------------------

class ResourceCreate(BaseModel):
    """Request body for POST /resources."""
    title: str = Field(..., min_length=1, max_length=200)
    category: str = Field(
        ...,
        min_length=1,
        max_length=50,
        description="资源分类，如 phonics/word_card/recommended/song/video 等",
    )
    pdf_url: str = Field("", min_length=0, description="逻辑存储路径，如 resources/phonics/xxx.pdf; 上传后自动填充")
    sort_order: int = Field(0, ge=0)
    is_active: bool = True


class ResourceUpdate(BaseModel):
    """Request body for PUT /resources/{id}. All optional."""
    title: str | None = Field(None, min_length=1, max_length=200)
    category: str | None = Field(None, min_length=1, max_length=50)
    sort_order: int | None = Field(None, ge=0)
    is_active: bool | None = None


class ResourceOut(BaseModel):
    """Resource returned in list views."""
    id: UUID
    title: str
    category: str
    pdf_url: str | None = None
    sort_order: int = 0
    is_active: bool = True
    created_at: datetime
    updated_at: datetime


class ResourceDetail(ResourceOut):
    """Resource detail with download URL."""
    pdf_url: str | None = None  # logical storage path (admin sees raw, user sees download)
    signed_pdf_url: str | None = None  # download endpoint path for parent/teacher


class PaginatedResources(BaseModel):
    """Paginated list response."""
    items: list[ResourceOut]
    total: int
    page: int
    page_size: int


# ---------------------------------------------------------------------------
# Generic PDF upload — admin
# ---------------------------------------------------------------------------

class UploadPdfOut(BaseModel):
    """Response for POST /upload/pdf — generic PDF upload."""
    storage_path: str = Field(..., description="逻辑存储路径 (如 uploads/xxx.pdf)")
    url: str = Field(..., description="同 storage_path，绑定到 resource.pdf_url")
